### PR TITLE
Integrate Hyperbus peripheral

### PR DESCRIPTION
**Pull request by sermazz**
This PR:
- integrates the Hyperbus peripheral in the Chimera top RTL
- integrates the HyperRAM VIP in the Chimera testbench
- adds an addressability test for the Hyperbus and adds it to the CI

- [ ] Config registers are still to be added https://github.com/pulp-platform/chimera/issues/43
- [ ] What address range should we use for the HyperRAM? https://github.com/pulp-platform/chimera/blob/smazzola/hyperbus/hw/chimera_pkg.sv#L128-L130

### CODE REVIEW

File: hw/chimera_top_wrapper.sv [hunk 1]
```diff
@@ -349,14 +363,17 @@ module chimera_top_wrapper
     .isolate_i        (pmu_iso_en_clusters_i),
     .isolate_o        (pmu_iso_ack_clusters_o)
   );
+  // Generate indices and get maps for all ports
+  localparam axi_in_t AxiIn = gen_axi_in(Cfg);
+  localparam axi_out_t AxiOut = gen_axi_out(Cfg);
```
Comment (Lore0599) [inline hw/chimera_top_wrapper.sv:368]: I think that the argument of these functions should be of type `cheshire_cfg_t`.

File: sw/tests/testHyperbusAddr.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#include <soc_addr_map.h>
+#include <stdint.h>
+
+#define HYPER_BASE 0x50000000
+#define TESTVAL (uint32_t)0x1234ABCD
+
+int main() {
+    volatile uint32_t *hyperMemPtr = (volatile uint32_t *)HYPER_BASE;
+    volatile uint32_t result;
+    volatile uint32_t golden = TESTVAL;
```
Comment (Lore0599) [inline sw/tests/testHyperbusAddr.c:18]: Is the variable `golden` used somewhere?

File: hw/chimera_pkg.sv [hunk 3]
```diff
@@ -117,13 +123,28 @@ ExtClusters
   localparam byte_bt MemIslNumWideBanks = 2;
   localparam shrt_bt MemIslWordsPerBank = 1024;
 
+  // Hyperbus
+  localparam byte_bt HyperbusIdx = MemIslandIdx + 1;
+  localparam doub_bt HyperbusRegionStart = 64'h5000_0000;
+  //TODO(smazzola): Correct size of HyperRAM?
```
Comment (Lore0599) [inline hw/chimera_pkg.sv:129]: We should discuss with the others about this point.

File: hw/chimera_pkg.sv [hunk 4]
```diff
@@ -117,13 +123,28 @@ ExtClusters
...
+  localparam doub_bt HyperbusRegionEnd = HyperbusRegionStart + 64'h0800_0000;
+
+  localparam int unsigned LogDepth = 3;
+  localparam int unsigned SyncStages = 3;
+
+  // Narrow patameters
+  localparam int unsigned AxiNarrowAddrWidth = 32;
+  localparam int unsigned AxiNarrowDataWidth = 32;
+  localparam int unsigned AxiNarrowStrobe = AxiNarrowDataWidth / 8;
+
```
Comment (Lore0599) [inline hw/chimera_pkg.sv:139]: Can we use the parameters already defined into the `gen_chimera_cfg` function? For example `cfg.AxiDataWidth` instead of `AxiNarrowAddrWidth`?

**Review by Lore0599 — COMMENTED** [review]
Overall LGTM

**Comment by sermazz** [comment]
@Lore0599 @arpansur @Scheremo @adimauro-iis 
CI is finally passing after the fixes on `devel`, we should now define the range we need for HyperRAM (https://github.com/pulp-platform/chimera/pull/47/files#diff-7310782802ac0d64363e8ceb9e90e54a5c0c858fc1e63688d48ad6964dd3fb4bR131). And also have a plan for https://github.com/pulp-platform/chimera/issues/43.
If these are not blocking, we can merge.

**Comment by Scheremo** [comment]
> @Lore0599 @arpansur @Scheremo @adimauro-iis CI is finally passing after the fixes on `devel`, we should now define the range we need for HyperRAM (https://github.com/pulp-platform/chimera/pull/47/files#diff-7310782802ac0d64363e8ceb9e90e54a5c0c858fc1e63688d48ad6964dd3fb4bR131). And also have a plan for #43. If these are not blocking, we can merge.

I think both are blocking but trivial to resolve. You can extend the register memory map range (0x3000_0000 - 0x4000_0000) contiguously with the config registers and map the hyperbus memory "somewhere reasonable" (why not 0x5000_0000?)

File: bender.mk [hunk 5]
```diff
@@ -8,4 +8,4 @@
 COMMON_TARGS ?=
 COMMON_TARGS += -t snitch_cluster -t cv32a6_convolve -t cva6 -t rtl
 
-SIM_TARGS = -t test -t sim
+SIM_TARGS = $(COMMON_TARGS) -t test -t sim
```
Comment (Lore0599) [inline bender.mk:11]: Why do you add here $(COMMON_TARGS)? Isn't the variable already used in `sim.mk`?

File: target/sim/sim.mk [hunk 6]
```diff
@@ -3,23 +3,46 @@
...
+CHIM_VLOG_ARGS += +define+HYP0_PRELOAD_MEM_FILE=\"$(HYP0_PRELOAD_MEM_FILE)\"
+CHIM_VLOG_ARGS += +define+HYP1_PRELOAD_MEM_FILE=\"$(HYP1_PRELOAD_MEM_FILE)\"
+# this path should be kept relative to the vsim directory to avoid CI issues:
+# an absolute path produce inter-CI-runner file accesses
+CHIM_VLOG_ARGS += +define+PATH_TO_HYP_SDF=\"../models/s27ks0641/s27ks0641.sdf\"
 
+# Generate vsim compilation script
 $(CHIM_SIM_DIR)/vsim/compile.tcl: chs-hw-init snitch-hw-init
-	@bender script vsim $(COMMON_TARGS) $(SIM_TARGS) --vlog-arg="$(VLOG_ARGS)"> $@
+	@bender script vsim $(SIM_TARGS) --vlog-arg="$(CHIM_VLOG_ARGS)" > $@
```
Comment (Lore0599) [inline target/sim/sim.mk:38]: Why did you remove $(COMMON_TARGS)?

File: target/sim/src/vip_chimera_soc.sv [hunk 7]
```diff
@@ -554,35 +567,105 @@ module vip_chimera_soc
...
+      Hyp1UserPreloadMemFile
+  };
+
+  for (genvar i = 0; i < HypNumPhys; i++) begin : hyperrams
+    for (genvar j = 0; j < HypNumChips; j++) begin : chips
+      s27ks0641 #(
+        .UserPreload  (HypUserPreload),
+        .mem_file_name(HypUserPreloadMemFiles[i]),
+        .TimingModel  ("S27KS0641DPBHI020")
+      ) dut (
```
Comment (Lore0599) [inline target/sim/src/vip_chimera_soc.sv:585]: The Chimera instance inside `fixture_chimera_soc.sv` is already called DUT.

File: bender.mk (hunk 5 above)
Comment (sermazz) [reply bender.mk:11]: [This also answers https://github.com/pulp-platform/chimera/pull/47#discussion_r1808717677]
Just a small modification to make (in my opinion) the flow clearer. At some point we'll have many targets in the `bender.mk`, e.g. `SIM_TARGS`, `ASIC_TARGS`, `FPGA_XILING_TARGS`, `FPGA_whatever_TARGS`, etc...
So I explicitly add `$(COMMON_TARGS)` (which is common to all of them) within all those other TARGS variables.
Previously, it was left to whoever wrote the independent target `.mk` (e.g., `sim.mk`, `asic.mk`, ...) to use both variables, as in: `bender script vsim $(COMMON_TARGS) $(SIM_TARGS) [...]`.
But ultimately it's the same argument for `bender` so we might as well use this clearer form.

Comment (Lore0599) [reply bender.mk:11]: make sense, thanks

File: hw/chimera_pkg.sv [hunk 8]
```diff
@@ -118,13 +124,23 @@ ExtClusters
   localparam byte_bt MemIslNumWideBanks = 2;
   localparam shrt_bt MemIslWordsPerBank = 1024;
 
+  // Hyperbus
+  localparam byte_bt HyperbusIdx = MemIslandIdx + 1;
+  localparam doub_bt HyperbusRegionStart = 64'h5000_0000;
+  //TODO(smazzola): Correct size of HyperRAM?
+  localparam doub_bt HyperbusRegionEnd = HyperbusRegionStart + 64'h0800_0000;
```
Comment (Scheremo) [inline hw/chimera_pkg.sv:131]: I think after our discussion last week, we can extend the region size of 64'h1000_0000

File: hw/chimera_top_wrapper.sv [hunk 9]
```diff
@@ -372,4 +389,160 @@ module chimera_top_wrapper
...
+    .axi_r_chan_t    (axi_slv_r_chan_t),
+    .axi_aw_chan_t   (axi_slv_aw_chan_t),
+    .RegAddrWidth    (ChsCfg.AddrWidth),
+    .RegDataWidth    (ChsCfg.AxiDataWidth),
+    .reg_req_t       (reg_req_t),
+    .reg_rsp_t       (reg_rsp_t),
+    .RxFifoLogDepth  (32'd2),
+    .TxFifoLogDepth  (32'd2),
+    .RstChipBase     (ChsCfg.LlcOutRegionStart),
+    .RstChipSpace    (HypNumPhys * HypNumChips * 'h800_0000),
```
Comment (Scheremo) [inline hw/chimera_top_wrapper.sv:500]: Not sure this is correct, can you comment?

File: target/sim/src/vip_chimera_soc.sv (hunk 7 above)
Comment (Scheremo) [reply target/sim/src/vip_chimera_soc.sv:585]: I think it's fine for now.

**Review by Scheremo — COMMENTED** [review]
Looks mostly fine to me, some comments sprinkled throughout. Is there some consideration behind having two PHYs? I understand two endpoints (HyperRAM and HyperFlash), but that should only necessitate another CS pin.

**Comment by sermazz** [comment]
Should be ready to merge.
- I integrated Morit's fixes from https://github.com/pulp-platform/chimera/commit/b67f438fccff22205e72bedda536a498b590ee10 and further fixed some small stuff
- Fixed the linting
- Connected the hyperbus peripheral to the regbus by instantiating another external rebus port in Cheshire

You can check the last commits that I added. Let me know if there's any further feedback :)

**Comment by Scheremo** [comment]
> Should be ready to merge.
> 
> * I integrated Morit's fixes from [b67f438](https://github.com/pulp-platform/chimera/commit/b67f438fccff22205e72bedda536a498b590ee10) and further fixed some small stuff
> * Fixed the linting
> * Connected the hyperbus peripheral to the regbus by instantiating another external regbus port in Cheshire
> 
> You can check the last commits that I added. Let me know if there's any further feedback :)

Hey Sergio, thanks for the effort! All my comments are satisfied; we might want to add a few more large-scale tests, but the RTL looks mergeable to me.

**Review by Scheremo — APPROVED** [review]
LGTM!
